From 5c755ed5e0ebfa7b87dd6bb49a4f5214af4d4b18 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <josh@wilsdon.ca>
Date: Wed, 13 Feb 2019 13:49:08 -0800
Subject: [PATCH] CNAPI-370 CommandExecute method needs configurable timeout
 Reviewed by: Orlando Vazquez <orlando@joyent.com> Reviewed by: Trent Mick
 <trentm@gmail.com> Approved by: Orlando Vazquez <orlando@joyent.com> Approved
 by: Trent Mick <trentm@gmail.com>

---
 docs/index.md            | 13 +++++++------
 lib/endpoints/servers.js | 10 ++++++++++
 package.json             |  2 +-
 3 files changed, 18 insertions(+), 7 deletions(-)

diff --git a/docs/index.md b/docs/index.md
index d5ea770..210f13b 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -1119,12 +1119,13 @@ executed non-zero a 500 error will be returned.
 
 ### Inputs
 
-| Param  | Type    | Description                                                                |
-| ------ | ------- | -------------------------------------------------------------------------- |
-| args   | Array   | Array containing arguments to be passed in to command                      |
-| env    | Object  | Object containing environment variables to be passed in                    |
-| script | String  | Script to be executed. Must have a shebang line                            |
-| json   | Boolean | Whether to return results as JSON instead of just stdout (default = false) |
+| Param   | Type    | Description                                                                                                                                                  |
+| ------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| args    | Array   | Array containing arguments to be passed in to command                                                                                                        |
+| env     | Object  | Object containing environment variables to be passed in                                                                                                      |
+| script  | String  | Script to be executed. Must have a shebang line                                                                                                              |
+| json    | Boolean | Whether to return results as JSON instead of just stdout (default = false)                                                                                   |
+| timeout | Integer | Number of ms to wait for command completion before killing task and returning (only supported when using cn-agent, See: FEATURE_USE_CNAGENT_COMMAND_EXECUTE) |
 
 
 ### Responses
diff --git a/lib/endpoints/servers.js b/lib/endpoints/servers.js
index 9c8ff31..4e81cf4 100644
--- a/lib/endpoints/servers.js
+++ b/lib/endpoints/servers.js
@@ -1136,6 +1136,7 @@ Server.ensureImage = function handlerServerEnsureImage(req, res, next) {
  * @param {Object} env Object containing environment variables to be passed in
  * @param {String} script Script to be executed. Must have a shebang line
  * @param {Boolean} json Whether to return results as JSON instead of just stdout (default = false)
+ * @param {Integer} timeout Number of ms to wait for command completion before killing task and returning (only supported when using cn-agent, See: FEATURE_USE_CNAGENT_COMMAND_EXECUTE)
  *
  * @response 404 None No such server
  * @response 500 None Error occurred executing script
@@ -1218,6 +1219,8 @@ Server.execute = function handlerCommandExecute(req, res, next) {
             return;
         }
     ]}, function sendRequest(err) {
+        var msg;
+
         if (err) {
             next(err);
             return;
@@ -1225,6 +1228,13 @@ Server.execute = function handlerCommandExecute(req, res, next) {
 
         if (handleWithUr) {
             // Ur's execute functionality
+            if (req.params.hasOwnProperty('timeout')) {
+                msg = 'timeout only supported when using cn-agent (See: ' +
+                    'FEATURE_USE_CNAGENT_COMMAND_EXECUTE)';
+                self.log.error(msg);
+                next(new restify.InvalidArgumentError(msg));
+                return;
+            }
             ur.execute(req, res, next);
         } else {
             // cn-agent's command_execute task
diff --git a/package.json b/package.json
index 0aeb2dc..3f3a6c7 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.19.2",
+  "version": "1.19.3",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

